extensions/sql-database-projects/src/models/project.ts

```diff
@@ -323,6 +328,23 @@ export class Project {
 		}
 	}
 
+	private databaseReferenceExists(entry: DatabaseReferenceProjectEntry): boolean {
+		const isSystemDatabaseProjectEntry = (<SystemDatabaseReferenceProjectEntry>entry).ssdtUri;
+		const entryFsPath = isSystemDatabaseProjectEntry ? entry.fsUri.fsPath.substring(1) : entry.fsUri.fsPath;
+
+		for (let r = 0; r < this.projFileXmlDoc.documentElement.getElementsByTagName(constants.ArtifactReference).length; r++) {
```

Comment:
Why not check against `this.databaseReferences.find(something)` instead of the xml?

Reply:
SSDT allows adding a reference with the same dacpac name but different filepath. this.databaseReferences only has the name of the dacpacs right now.

Reply:
You should at least turn this into a `.this.projFileXmlDoc.documentElement.getElementsByTagName(constants.ArtifactReference).find(ele => ...)`

Right now you're calling `getElementsByTagName` once at the beginning and then 2 times for every element found - using find will reduce that to one call at the beginning.

Reply:
there isn't find() on NodeList, but I changed it to just call getElementsByTagName once at the beginning

Reply:
Is there a reason it makes more sense to check the XML over adding the filepath info to `this.databaseReferences`?  I feel like we should keep talking to the XML should be kept to the serialization/deserialization methods as much as possible since that logic is more fragile and probably going to move out of this class.

Reply:
Sure makes sense. I changed this.databaseReferences to be DatabaseReferenceProjectEntry[] instead of string[] and removed the XML checking from addDatabaseReferenceExists()